### DISCUSSION TITLE
`response_model_exclude_unset` with `default_factory` is ignored?

### DISCUSSION THREAD

**Discussion by jkittner** [fastapi/fastapi#13112] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi.testclient import TestClient
from pydantic import BaseModel, Field
from typing import Generic, TypeVar

T = TypeVar('T')

def timestamp() -> int:
    return 1735326936124

class Response(BaseModel, Generic[T]):
    data: T
    timestamp: int = Field(default_factory=timestamp)

class Model(BaseModel):
    name: str
    sth_else: str | None = Field(None)

app = FastAPI()

@app.get( "/", response_model=Response[Model], response_model_exclude_unset=True)
async def route():
    return Response(data={"name": "test"})

# test case:

def client():
    with TestClient(app) as client:
        yield client

client = TestClient(app)

def test_response():
    response = client.get("/")
    assert response.status_code == 200
    assert response.json() == {"data": {"name": "test"}, "timestamp": 1735326936124}
```

### Description

With the above code I expect the timestamp field to get populated so the response to `/` would look like this: 

```json
{"data":{"name":"test"}, "timestamp":1735326936124}
```

but instead it is ignored:

```json
{"data":{"name":"test"}}
```

It works fine if I remove `response_model_exclude_unset=True` but I don't want to show `sth_else` as `null`, so I need this set!

I checked via `breakpoint` - the default_factory is actually executed, but the result not used in the response?

This used to work - I am not sure what changed?

Maybe someone can point me at what changed and I need to fix?

### Operating System

Linux

### FastAPI Version

0.115.6

### Pydantic Version

2.10.4

### Python Version

Python 3.12.8

**Comment by luzzodev** [comment, marked as answer]
Hello @jkittner,

the option `response_model_exclude_unset` will exclude all fields that are not explicitly set. Using default or default_factory is not considered an explicit set. In your case you can use exclude_none instead of exclude_unset.

**Comment by jkittner** [comment]
Mhm, I did dig a little deeper and found that it broke with pydantic 2.10 - it was fine in pydantic 2.9 though. One issue with `response_model_exclude_none` is shown by this test case:


```python
@app.get( "/", response_model=Response[list[Model]], response_model_exclude_none=True)
async def route():
    return Response(
        data=[
            {"name": "test", "sth_else": None},
            {"name": "test", "sth_else": "test"}
        ]
    )

# test case:

def client():
    with TestClient(app) as client:
        yield client

client = TestClient(app)

def test_response():
    response = client.get("/")
    assert response.status_code == 200
    assert response.json() == {
        "data": [
            {"name": "test", "sth_else": None},
            {"name": "test", "sth_else": "test"},
        ],
        "timestamp": 1735326936124
    }
```

```
  Full diff:
    {
        'data': [
            {
                'name': 'test',
  -             'sth_else': None,
            },
            {
                'name': 'test',
                'sth_else': 'test',
            },
        ],
        'timestamp': 1735326936124,
    }
```

Even though it was **exclicitly** set to `None`, it is removed, hence object in the array become inconsistent, which is never a good thing!

So that does not give me the old behavior back...

**Reply by luzzodev** [reply]
`response_model_exclude_none` will exclude all values that are None. There is no distinction between set it to None implicit or explicit.
Said that, your test case is wrong and you should expect

``` python 
def test_response():
    response = client.get("/")
    assert response.status_code == 200
    assert response.json() == {
        "data": [
            {"name": "test"},
            {"name": "test", "sth_else": "test"},
        ],
        "timestamp": 1735326936124
    }
```
that seems to be your expected case in the first comment.
Am i missing something?

**Comment by jkittner** [comment]
I bisected pydantic and came out here: https://github.com/pydantic/pydantic/commit/f3092b337e703ef1df07259ba691a94687ef1b72

The weird thing is, I am not able to reproduce this with just pydantic. No matter what version of pydantic I use, it always has the "new" behavior with the timestamp missing...

**Reply by luzzodev** [reply]
Like i said before, exclude_unset will exclude all values that are not explicitly set. Using default or default_factory is not considered an explicit set and this behavior is the same in all version of pydantic 2.x so maybe that could be the reason why you are not able to reproduce only with pydantic.

**Reply by jkittner** [reply]
Mhm, it must be a combination of both, fastapi and pydantic.
Try the example from the question with pydantic 2.9 and 2.10 **with fastapi** and see that there is a different behavior just by changing the pydantic version.
I am trying to investigate what exactly changed and what I need to do to get the previous behavior, which was exactly what I wanted and wrote all sorts of tests for.

**Reply by luzzodev** [reply]
I can give you some hints to reproduce your issue without fastapi. You should use TypeAdapters ( comes from pydantic) to validate and serialize the response like fastapi does. 
You can check the code https://github.com/fastapi/fastapi/blob/741b95a45423012b660056506312b781907d91d7/fastapi/_compat.py#L137

To be honest it seems more like a fix then a bug from pydantic, anyway i think you should discuss about it there.
Hope this helps.
